sycl/doc/extensions/proposed/sycl_ext_oneapi_usm_malloc_properties.asciidoc

```diff
@@ -0,0 +1,1615 @@
+`SYCL_EXT_ONEAPI_USM_MALLOC_PROPERTIES` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features
+the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version
+|===
+
+=== Orientation
+
+The core SYCL specification lists eight functions in each of the following four tables:
+
+- Table 103 "Device USM Allocation Functions", of Section 4.8.3.2 "Device allocation functions" +
+- Table 104 "Host USM Allocation Functions", of Section 4.8.3.3 "Host allocation functions" +
+- Table 105 "Shared USM Allocation Functions", of Section 4.8.3.4 "Shared allocation functions" +
+- Table 106 "Parameterized USM Allocation Functions", of Section 4.8.3.5 "Parameterized allocation functions"
+
+This extension introduces a new function for each function listed in the above tables of the core SYCL specification.
+This extension also adds USM memory allocation functions with `properties` support that require the `sycl::ext::oneapi::experimental::usm_kind` property,
+and these do not correspond to any functions in the core SYCL specification.
+All USM memory allocation functions introduced by this extension are listed explicitly in Section <<section.usm.allocs>>.
+
+[NOTE]
+====
+The USM memory allocation functions defined in the core SYCL specification can be used in the same program as the USM memory allocation functions with `properties` support defined in this extension.
+The new functions are distinguished by their names having the `annotated_` prefix.
+====
+
+[[section.usm.allocs]]
+=== USM Memory Allocation Functions with properties Support
+
+The following five tables list all functions introduced by this extension.
+
+[NOTE]
+====
+All functions in the following five tables belong to the `sycl::ext::oneapi::experimental` namespace. The
+namespace is omitted to save space.
+====
+
+[[section.usm.device.allocs]]
+==== Device USM allocation functions with properties support
+
+[[table.usm.device.allocs]]
+.Device USM Allocation Functions with properties Support
+[width="100%",options="header",separator="@",cols="65%,35%"]
+|====
+@ Function @ Description
+a@
+[source]
+----
+template<typename propertyListA, typename propertyListB>
+annotated_ptr<void, propertyListB>
+annotated_malloc_device(
+   size_t numBytes,
+   const device& syclDevice,
+   const context& syclContext,
+   const propertyListA &propList = properties{})
+----
+a@ Returns an `annotated_ptr` containing a raw pointer to the newly allocated memory on `syclDevice` on
+success.  The allocation size is specified in bytes.  This memory is not
+accessible on the host.  Memory allocated by `annotated_malloc_device`
+must be deallocated with `sycl::free` to avoid memory leaks.
+On failure, the raw pointer of the returned `annotated_ptr` will be `nullptr`.
```

Comment:
I guess it is not possible to return `nullptr` here which is a C++ keyword and obviously is of type https://en.cppreference.com/w/cpp/types/nullptr_t
Perhaps you could have a `using nullptr_t = ...` and `constexpr static nullptr_v = ...` inside your annotated pointer class for this?

Reply:
We are discussing adding a comparison operator to `annotated_ptr`, e.g.:

` bool operator == (const nullptr_t &other) const { return this->ptr == nullptr; }`
which would allow us to write

`
annotated_ptr x = malloc_device_annotated<int>(N, q);
if (x == nullptr) {…}
`

Reply:
Great idea.
You could also have a conversion-to-bool operator, if you do not have one already.

Reply:
The annotated_ptr spec currently lists a conversion to bool operator:

`explicit operator bool() const noexcept;`
Returns false if the underlying pointer is null, returns true otherwise.